Reenable Python 3.12 wheel building

djhoese opened the pull request:
I'm working on building Python 3.12 wheels for some of my packages that depend on shapely. It'd be great to have a shapely wheel for Python 3.12.

Side note: I think in pyproject.toml you're supposed to be able to just say "numpy" from 1.25 or 1.26 on. Of course I can't find it now, but I believe it is supposed to be bundled with backwards compatible APIs...I think.

djhoese commented:
As a library maintainer, I'm also curious why you make the individual macos wheels and also the universal2 wheel?

caspervdw commented:
> As a library maintainer, I'm also curious why you make the individual macos wheels and also the universal2 wheel?

I am not sure actually. Looking at numpy, they *stopped* the universal2 wheel since version 1.22. https://github.com/numpy/numpy/issues/21233

Maybe we should too? Something for a different PR.

Either way, good to start testing and distribution for Python 3.12. Could you also append the test.yml?

djhoese commented:
I added it...I think, but I wasn't really sure what pattern to follow. Numpy 1.26.0 was just released so I set numpy to that. I also added a flag so Python 3.12 pre-releases can get used since 3.12 doesn't have a stable release out yet.

djhoese commented:
All the failures seem to be from this deprecation warning:

```
 >   EPOCH = datetime.datetime.utcfromtimestamp(0)
E   DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
```

djhoese commented:
Oh this is actually coming from dateutil which seems decently behind in stable releases supporting modern python versions. Maybe add this warning to the ignored warnings?

djhoese commented:
Ah it's already fixed, they just need a release: https://github.com/dateutil/dateutil/commit/f2293200747fb03d56c6c5997bfebeabe703576f

djhoese commented:
Ok, this is getting worse and worse. There was the deprecation warning about dateutil so I was going to filter it in setup.cfg, but pytest config `filterwarnings` are not the priority anymore so the `-W error` in the CI execution were still making it error out. So I removed the `-W error` and added `error` to the config...but now there is a pyproj warning that seems like something that should really be fixed...but also isn't specific to Python 3.12 from what I can tell?

mwtoews reviewed (approved):
Thanks @djhoese, this looks good! I'll be interested in looking at the macos and numpy questions in the next few months. Some of the cibuildwheel configuration is different than numpy as shapely covers a few earlier versions of Python.

mwtoews commented:
I've pushed some changes to use the "2023" matrix item for as the actual release year for Python 3.12, and also use the "dev" matrix item for Python 3.12.

djhoese commented:
> I've pushed some changes to use the "2023" matrix item for as the actual release year for Python 3.12, and also use the "dev" matrix item for Python 3.12.

Sounds good.

>  I'll be interested in looking at the macos and numpy questions in the next few months. Some of the cibuildwheel configuration is different than numpy as shapely covers a few earlier versions of Python.

This comes from a place of being a library maintainer myself, what's the reason for supporting older versions of Python?

mwtoews commented:
> This comes from a place of being a library maintainer myself and curiousity, what's the reason for supporting older versions of Python?

Fair question. The current status quo is to provide built wheels for the 4 to 5 most recent Python versions. Unifying around [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) and/or [SPEC 0](https://scientific-python.org/specs/spec-0000/) is an open discussion. There is now a good argument for adopting these policies, as NumPy is was brought as a requirement since Shapely 2.0. (Also, my personal observation is that folks are quicker to upgrade their Python environments now than they were, say, 5 years ago when there was an awkward mix of Python 2 and 3.)

djhoese commented:
>  (Also, my personal observation is that folks are quicker to upgrade their Python environments now than they were, say, 5 years ago when there was an awkward mix of Python 2 and 3.)

Ah ok. I thought you were saying much older versions (3.7 or older). Yeah I agree with this observation. I also like the excuse of "oh you want the new features in the library, you'll need to update your version of python, please use installation solution X, Y, or Z".

djhoese commented:
So the obvious question is, what is going to be the next released version and when do you expect to be able to make it?

mwtoews commented:
Appreciate your patience @djhoese as I'd like to see movement on #1852 too. Shapely 2.1 is not ready yet, but a micro release from `maint-2.0` should be ready. Xref #1898
